packages/core/src/routes/sign-in-experience/custom-ui-assets/index.ts

```diff
@@ -0,0 +1,100 @@
+import { getTenantId } from '#src/utils/tenant.js';
+
+import { type ManagementApiRouter, type RouterInitArgs } from '../../types.js';
+
+export default function customUiAssetsRoutes<T extends ManagementApiRouter>(
+  ...[router]: RouterInitArgs<T>
+) {
+  // TODO: Remove
+  if (!EnvSet.values.isDevFeaturesEnabled) {
+    return;
+  }
+
+  router.post(
+    '/sign-in-exp/default/custom-ui-assets',
+    koaGuard({
+      files: object({
+        file: uploadFileGuard.array().min(1),
+      }),
+      response: z.object({
+        customUiAssetId: z.string(),
+      }),
+      status: [200, 400, 500],
+    }),
+    async (ctx, next) => {
+      const { file: bodyFiles } = ctx.guard.files;
+      const file = bodyFiles[0];
+
+      assertThat(file, 'guard.invalid_input');
+      assertThat(file.size <= maxUploadFileSize, 'guard.file_size_exceeded');
+      assertThat(file.mimetype === 'application/zip', 'guard.mime_type_not_allowed');
+
+      const { experienceZipsProviderConfig } = SystemContext.shared;
+      assertThat(
+        experienceZipsProviderConfig?.provider === 'AzureStorage',
+        'storage.not_configured'
+      );
+      const { connectionString, container } = experienceZipsProviderConfig;
+
+      const { uploadFile, downloadFile, isFileExisted } = buildAzureStorage(
+        connectionString,
+        container
+      );
+
+      const [tenantId] = await getTenantId(ctx.URL);
+      assertThat(tenantId, 'guard.can_not_get_tenant_id');
+
+      const customUiAssetId = generateStandardId(8);
+      const objectKey = `${tenantId}/${customUiAssetId}/${file.originalFilename}`;
+      const errorLogObjectKey = `${tenantId}/${customUiAssetId}/error.log`;
+
+      try {
+        // Upload the zip file to `experience-zips` container, in which a blob trigger is configured,
+        // and an azure function will be executed automatically to unzip the file on blob received.
+        await uploadFile(await readFile(file.filepath), objectKey, {
+          contentType: file.mimetype,
+        });
+
+        // Check the unzipping status every 500ms until the request is timed out
+        await new Promise((resolve, reject) => {
+          const intervalId = setInterval(async () => {
+            const zipExisted = await isFileExisted(objectKey);
+            const hasError = await isFileExisted(errorLogObjectKey);
+            // If there is no zip file and no error log either, consider the unzipping as successful
+            if (!zipExisted && !hasError) {
+              clearInterval(intervalId);
+              resolve(true);
+            }
+            // If there is no zip file but an error log exists, consider the unzipping as failed
+            if (hasError) {
+              clearInterval(intervalId);
+              const errorLog = await downloadFile(errorLogObjectKey);
+              // Log the detailed error message
+              getConsoleLogFromContext(ctx).error(errorLog.readableStreamBody?.read());
```

Comment:
we can show the detail message in the response body

Reply:
updated